README.md
# ShowEQ


ChangeLog
BlueAdept (9/20/18)
-------------------
- Updated version to 5.13.10.72
- Support EQ patch 9/19/18
- Updated Opcodes (Newby)
- Updated Structs (Newby)

BlueAdept (8/16/18)
-------------------
- Updated version to 5.13.10.71
- Support EQ patch 8/15/18
- Updated Opcodes (Newby)
- Updated Structs (Newby)

BlueAdept (7/18/18)
-------------------
- Updated version to 5.13.10.70
- Support EQ patch 7/18/18
- Updated Opcodes (Newby)
- Updated Structs (Newby)
- Added a "fflush (stdout);" to terminal.cpp
- Bumped version to .70 due to an error in .69

BlueAdept (7/06/18)
-------------------
- Updated version to 5.13.10.68
- Patch to fix the Log Spawns

BlueAdept (6/23/18)
-------------------
- Updated version to 5.13.10.67
- Support EQ patch 6/20/18
- Updated Opcodes (Newby)
- Updated Structs (Newby)

BlueAdept (5/18/18)
-------------------
- Updated version to 5.13.10.66
- Support EQ patch 5/16/18
- Updated Opcodes (Newby)
- Updated Structs (Newby)

BlueAdept (4/19/18)
-------------------
- Updated version to 5.13.10.65
- Support EQ patch 4/20/18
- Updated Opcodes (Newby)
- Updated Structs (Newby)

BlueAdept (3/14/18)
-------------------
- Updated version to 5.13.10.64
- Support EQ patch 3/14/18
- Updated Opcodes (Newby)
- Updated Structs (Newby)

BlueAdept (2/22/18)
-------------------
- Updated version to 5.13.10.63
- Support EQ patch 2/21/18
- Updated Opcodes (Newby)
- Updated Structs (Newby)

BlueAdept (1/19/18)
-------------------
- Updated version to 5.13.10.62
- Support EQ patch 1/17/18
- Updated Opcodes (Newby)
- Updated Structs (Newby)

BlueAdept (1/09/18)
-------------------
- Updated version to 5.13.10.61
- Support EQ patch 1/05/18
- Updated Opcodes (Newby)
- Updated Structs (Newby)

BlueAdept (12/23/17)
-------------------
- Updated version to 5.13.10.60
- Support EQ patch 12/19/17
- Updated Opcodes (Newby)
- Updated Structs (Newby)
- Updated zones.h (llaffer)
- Updated races.h (fransick)

BlueAdept (12/13/17)
-------------------
- Updated version to 5.13.10.59
- Support EQ patch 12/12/17
- Updated Opcodes (Newby)
- Updated Structs (Newby)

BlueAdept (11/23/17)
-------------------
- Updated version to 5.13.10.58
- Support EQ patch 11/21/17
- Updated Opcodes (Newby)
- Updated Structs (Newby)

BlueAdept (11/19/17)
-------------------
- Updated version to 5.13.10.57
- Support EQ patch 11/15/17
- Updated Opcodes (Newby)
- Updated Structs (Newby)

BlueAdept (10/21/17)
-------------------
- Updated version to 5.13.10.56
- Support EQ patch 10/19/17
- Updated Opcodes (Newby)
- Updated Structs (Newby)

BlueAdept (9/21/17)
-------------------
- Updated version to 5.13.10.55
- Support EQ patch 9/20/17
- Updated Opcodes (Newby)
- Updated Structs (Newby)

BlueAdept (8/19/17)
-------------------
- Updated version to 5.13.10.54
- Support EQ patch 8/16/17
- Updated Opcodes (Newby)
- Updated Structs (Newby)

BlueAdept (7/21/17)
-------------------
- Updated version to 5.13.10.53
- Support EQ patch 7/19/17
- Updated Opcodes (Newby)
- Updated Structs (Newby)

BlueAdept (6/30/17)
-------------------
- Updated version to 5.13.10.52
- Support EQ patch 6/21/17
- Updated Opcodes (Newby)
- Updated Structs (Newby)

BlueAdept (5/21/17)
-------------------
- Updated version to 5.13.10.51
- Support EQ patch 5/17/17
- Updated Opcodes (Newby)
- Updated Structs (Newby)


BlueAdept (4/28/17)
-------------------
- Updated version to 5.13.10.50
- Support EQ patch 4/26/17
- Updated Opcodes (Newby)
- Updated Structs (Newby)

BlueAdept (4/22/17)
-------------------
- Updated version to 5.13.10.49
- Support EQ patch 4/19/17
- Updated Opcodes (Newby)
- Updated Structs (Newby)

BlueAdept (3/18/17
-------------------
- Updated version to 5.13.10.48
- Support EQ patch 2/16/17
- Updated Opcodes (Newby)
- Updated Movement Structs (Newby)
- Updated Location Structs (Newby)

BlueAdept (2/25/17)
-------------------
- Updated version to 5.13.10.47
- Support EQ patch 2/15/17
- Updated Opcodes (Newby)
- Updated Structs (Newby)

BlueAdept (1/20/17)
-------------------
- Updated version to 5.13.10.46
- Support EQ patch 1/18/17
- Updated Opcodes (Newby)
- Updated position structs (Newby)

BlueAdept (12/18/16)
-------------------
- Updated version to 5.13.10.45
- Support EQ patch 12/14/16
- Updated Opcodes (Newby)
- Updated structs (Newby)
- Fixed run speed (showeq42)

BlueAdept (11/17/16)
-------------------
- Updated version to 5.13.10.44
- Support EQ patch 11/16/16
- Fixed guild information (Newby)

BlueAdept (11/17/16)
-------------------
- Updated version to 5.13.10.43
- Support EQ patch 11/16/16
- Updated Opcodes (Newby)
- Updated position structs (Newby)

BlueAdept (10/21/16)
-------------------
- Updated version to 5.13.10.42
- Support EQ patch 10/19/16
- Updated Opcodes (Newby)
- Updated position structs (Newby)


BlueAdept (09/24/16)
-------------------
- Updated version to 5.13.10.41
- Support EQ patch 09/21/16
- Updated Opcodes (Newby)
- Updated position structs (Newby)


BlueAdept (08/18/16)
-------------------
- Updated version to 5.13.10.40
- Support EQ patch 08/16/16
- Updated Opcodes (Newby)
- Updated position structs (Newby)

BlueAdept (07/15/16)
-------------------
- Updated version to 5.13.10.39
- Support EQ patch 07/13/16
- Updated Opcodes (Newby)
- Updated position structs (Newby)

BlueAdept (07/09/16)
--------------------
- Updated version to 5.13.10.38
- Support EQ patch 07/06/16
- Updated position structs (Newby)
- Updated Opcodes (Newby)

Newby (11/18/15)
-----------------
- Updated version to 5.13.10.37
- Support EQ patch 11/18/15
- Updated position structs
- Updated Opcodes

Newby (10/21/15)
-----------------
- Updated version to 5.13.10.36
- Support EQ patch 10/21/15
- Updated position structs
- Updated Opcodes

Newby (07/23/15)
-----------------
- Updated version to 5.13.10.35
- Support EQ patch 07/23/15
- Updated position structs
- Updated Opcodes

Fransick (05/28/15)
-----------------
- Updated version to 5.13.10.34
- Support EQ patch 05/28/15
- Updated position structs
- Minor struct changes

src/spawnlog.h
/*
 * spawnlog.h
 *
 *  ShowEQ Distributed under GPL
 *  http://www.sourceforge.net/projects/seq
 *
 *  Copyright 2000-2003 by the respective ShowEQ Developers
 *  Portions Copyright 2001-2003 Zaphod (dohpaz@users.sourceforge.net). 
 */

#ifndef _SPAWNLOG_H_
#define _SPAWNLOG_H_

#include <qobject.h>
#include "logger.h"
#include "everquest.h"

//----------------------------------------------------------------------
// forward declarations
class QDateTime;
class QTime;
class DateTimeMgr;
class Item;

//----------------------------------------------------------------------
// SpawnLog
class SpawnLog: public SEQLogger 
{
   Q_OBJECT

public:
    SpawnLog(DateTimeMgr* dateTimeMgr, const QString& filename);
    SpawnLog(DateTimeMgr* dateTimeMgr, FILE *fp);

public slots:
    void logNewZone(const QString& zone);
    void logZoneSpawns(const uint8_t* zspawns, size_t len);
    void logNewSpawn(const uint8_t* spawn);
    void logKilledSpawn(const Item* item, const Item* kitem, uint16_t kid);
    void logDeleteSpawn(const Item *spawn);
    void logNewSpawn(const Item *spawn);

protected:
    void logSpawnInfo(const char *type, const char *name, int id, int level, 
                      int x, int y, int z, 
                      const char *killer, int kid, int guildid);
    void logSpawnInfo(const char *type, const char *name, int id, int level, 
                      int x, int y, int z, 
		      const QDateTime& eqDate, const QTime& time, 
                      const char *killer, int kid, int guildid);
    int version;
    QString zoneShortName;
    DateTimeMgr* m_dateTimeMgr;
};

#endif // _SPAWNLOG_H_


src/spawnlog.cpp
/*
 * spawnlog.cpp
 *
 *  ShowEQ Distributed under GPL
 *  http://www.sourceforge.net/projects/seq
 *
 *  Copyright 2001-2003 by the respective ShowEQ Developers
 *  Portions Copyright 2001-2003,2007 Zaphod (dohpaz@users.sourceforge.net). 
 */

#include "spawn.h"
#include "datetimemgr.h"
#include "everquest.h"
#include "spawnlog.h"

SpawnLog::SpawnLog(DateTimeMgr* dateTimeMgr, const QString& fname)
  : SEQLogger(fname, NULL, "SpawnLog"),
    m_dateTimeMgr(dateTimeMgr)
{
    version = 4;
    zoneShortName = "unknown";
}

SpawnLog::SpawnLog(DateTimeMgr* dateTimeMgr, FILE *fp)
  : SEQLogger(fp, NULL, "SpawnLog"),
    m_dateTimeMgr(dateTimeMgr)
{
    version = 4;
    zoneShortName = "unknown";
}

inline void 
SpawnLog::logSpawnInfo(const char *type, const char *name, int id, int level,
                          int x, int y, int z, 
                          const char *killedBy, int kid, int guildid)
{
  const QDateTime& eqDate = m_dateTimeMgr->updatedDateTime();
  const QTime& time = QTime::currentTime(Qt::LocalTime);

  logSpawnInfo(type, name, id, level, x, y, z, 
	       eqDate, time, 
	       killedBy, kid, guildid);
}

void
SpawnLog::logSpawnInfo(const char *type, const char *name, int id, int level,
                          int x, int y, int z, 
			  const QDateTime& eqDate, const QTime& time,
                          const char *killedBy, int kid, int guildid)
{
  if (!open())
      return;
  
  const QDate& eqDateDate = eqDate.date();
  const QTime& eqDateTime = eqDate.time();
  
  outputf("%s:%s(%d):%d:%d,%d,%d:%02d.%02d.%02d:%d:%s:%02d.%02d.%02d.%02d.%04d:%s(%d):%d\n",
	  type,
	  name,
	  id,
	  level,
	  x,
	  y,
	  z,
	  time.hour(), time.minute(), time.second(),
	  version,
	  (const char*)zoneShortName,
	  eqDateTime.hour(),
	  eqDateTime.minute(),
	  eqDateDate.month(),
	  eqDateDate.day(),
	  eqDateDate.year(),
	  killedBy,
	  kid,
	  guildid);
  
  flush();
}

void 
SpawnLog::logZoneSpawns(const uint8_t* data, size_t len)
{
  const spawnStruct* zspawns = (const spawnStruct*)data;
  int spawndatasize = len / sizeof(spawnStruct);
  
  const QDateTime& eqDate = m_dateTimeMgr->updatedDateTime();
  const QTime& time = QTime::currentTime(Qt::LocalTime);
  
  for (int i = 0; i < spawndatasize; i++)
  {
    const spawnStruct& spawn = zspawns[i];
    logSpawnInfo("z",spawn.name,spawn.spawnId,spawn.level,
                 (spawn.x >> 3), (spawn.y >> 3), (spawn.z >> 3), 
		 eqDate, time, "", 0, spawn.guildID);
  }
}

void
SpawnLog::logNewSpawn(const uint8_t* data)
{
  const spawnStruct& spawn = *(const spawnStruct*)data;
  logSpawnInfo("+",spawn.name,spawn.spawnId,spawn.level,
	       (spawn.x >> 3), (spawn.y >> 3), (spawn.z >> 3), 
	       "", 0, spawn.guildID);
}

void
SpawnLog::logNewSpawn(const Item *item)
{
  if (item->type() != tSpawn)
    return;
  
  const Spawn* spawn = (const Spawn*)item;
  
  logSpawnInfo("+",(const char *)spawn->name(),spawn->id(),spawn->level(),
	       spawn->x(), spawn->y(), spawn->z(), "",0, spawn->guildID());
}

void
SpawnLog::logKilledSpawn(const Item *item, const Item* kitem, uint16_t kid)
{
  if (item == NULL)
    return;
  
  const Spawn* spawn = (const Spawn*)item;
  const Spawn* killer = (const Spawn*)kitem;
  
  logSpawnInfo("x",(const char *) spawn->name(),spawn->id(), spawn->level(), 
	       spawn->x(), spawn->y(), spawn->z(), 
	       killer ? (const char*)killer->name() : "unknown",
	       kid, spawn->guildID());
}

void
SpawnLog::logDeleteSpawn(const Item *item)
{
  if (item->type() != tSpawn)
    return;
  
  const Spawn* spawn = (const Spawn*)item;
  
  logSpawnInfo("-",(const char *)spawn->name(),spawn->id(),spawn->level(),
	       spawn->x(), spawn->y(), spawn->z(), "",0, spawn->guildID());
}

void
SpawnLog::logNewZone(const QString& zonename)
{
  if (!open())
      return;

  outputf("----------\nNEW ZONE: %s\n----------\n", (const char*)zonename);
  outputf(" :name(spawnID):Level:Xpos:Ypos:Zpos:H.m.s:Ver:Zone:eqHour.eqMinute.eqMonth.eqDay.eqYear:killedBy(spawnID)\n");
  flush();
  zoneShortName = zonename;
}

#ifndef QMAKEBUILD
#include "spawnlog.moc"
#endif



src/terminal.cpp
/*
 * terminal.cpp
 * 
 * ShowEQ Distributed under GPL
 * http://seq.sourceforge.net/
 *
 * Copyright 2003-2007 Zaphod (dohpaz@users.sourceforge.net)
 *
 */

#include "terminal.h"
#include "main.h"

//----------------------------------------------------------------------
// Constants

// ISO/ANSI/ECMA graphics codes, refer to ECMA-048 Section 8.3.117 SGR
//   available here http://www.ecma.ch/ecma1/STAND/ECMA-048.HTM
const char* defaultStyle = "\e[0m";
const char* typeStyles[] = 
  {
    "\e[1;32m",   // 0 - Guild
    NULL,         // 1
    "\e[0;36m",   // 2 - Group
    "\e[1;31m",   // 3 - Shout
    "\e[1;42m",   // 4 - Auction
    "\e[0;32m",   // 5 - OOC
    NULL,         // 6
    "\e[0;35m",   // 7 - Tell
    "\e[1;38m",   // 8 - Say
    NULL,         // 9
    NULL,         // 10
    "\e[5;31m",   // 11 - GM Say
    NULL,         // 12
    NULL,         // 13
    "\e[5;31m",   // 14 - GM Tell 
    "\e[1;36m",   // 15 - Raid
    NULL,         // 16 - Debug
    NULL,         // 17 - Info
    NULL,         // 18 - General
    NULL,         // 19 - Motd
    NULL,         // 20 - System
    NULL,         // 21 - Money
    NULL,         // 22 - Random
    NULL,         // 23 - Emote
    NULL,         // 24 - Time
    NULL,         // 25 - Spell
    NULL,         // 26 - Zone
    NULL,         // 27 - Inspect
    NULL,         // 28 - Player
    NULL,         // 29 - Consider
    "\e[1;37;45m",   // 30 - Alert
    "\e[1;37;41m",   // 31 - Danger
    "\e[1;37;43m",   // 32 - Caution
    "\e[1;37;46m",   // 33 - Hunt
    "\e[1;37;44m",   // 34 - Locate
  };

//----------------------------------------------------------------------
// Terminal
Terminal::Terminal(Messages* messages,
		   QObject* parent, const char* name)
  : QObject(parent, name),
    m_messages(messages),
    m_enabledTypes(0xFFFFFFFFFFFFFFFFULL),
    m_itemPattern("\022\\d(.{5}).{39}([^\022]+)\022"),
    m_dateTimeFormat("ddd M/d/yyyy h:mm:ss"),
    m_eqDateTimeFormat("ddd M/d/yyyy h:mm:ss"),
    m_displayType(true),
    m_displayDateTime(false),
    m_displayEQDateTime(false),
    m_useColor(true)
{
  const QString preferenceName = "Terminal";
  m_enabledTypes = pSEQPrefs->getPrefUInt64("EnabledTypes", preferenceName, 
					    m_enabledTypes);
  m_dateTimeFormat = pSEQPrefs->getPrefString("DateTimeFormat",
					      preferenceName, 
					      m_dateTimeFormat);
  m_eqDateTimeFormat = pSEQPrefs->getPrefString("EQDateTimeFormat",
						preferenceName,
						m_eqDateTimeFormat);
  m_displayType = pSEQPrefs->getPrefBool("DisplayType", preferenceName,
					 m_displayType);
  m_displayDateTime = pSEQPrefs->getPrefBool("DisplayDateTime",
					     preferenceName, 
					     m_displayDateTime);
  m_displayEQDateTime = pSEQPrefs->getPrefBool("DisplayEQDateTime",
					       preferenceName,
					       m_displayEQDateTime);
  m_useColor = pSEQPrefs->getPrefBool("UseColor", preferenceName,
				      m_useColor);

  // connect to the Messages signal(s)
  connect(m_messages, SIGNAL(newMessage(const MessageEntry&)),
	  this, SLOT(newMessage(const MessageEntry&)));
}

Terminal::~Terminal()
{
}

void Terminal::setEnabledTypes(uint64_t types) 
{ 
  m_enabledTypes = types; 
  pSEQPrefs->setPrefUInt64("EnabledTypes", "Terminal", m_enabledTypes);
}

void Terminal::setEnabledShowUserFilters(uint32_t filters)
{
  m_enabledShowUserFilters = filters;
  // save the new setting
  pSEQPrefs->setPrefUInt("EnabledShowUserFilters", "Terminal", 
			 m_enabledShowUserFilters);
}

void Terminal::setEnabledHideUserFilters(uint32_t filters)
{
  m_enabledHideUserFilters = filters;
  // save the new setting
  pSEQPrefs->setPrefUInt("EnabledHideUserFilters", "Terminal", 
			 m_enabledHideUserFilters);
}

void Terminal::setDateTimeForamt(const QString& dateTime)
{
  m_dateTimeFormat = dateTime;
  pSEQPrefs->setPrefString("DateTimeFormat", "Terminal", m_dateTimeFormat);
}

void Terminal::setEQDateTimeFormat(const QString& dateTime)
{
  m_eqDateTimeFormat = dateTime;
  pSEQPrefs->setPrefString("EQDateTimeFormat", "Terminal", m_eqDateTimeFormat);
}

void Terminal::setDisplayType(bool enable) 
{
  m_displayType = enable; 
  pSEQPrefs->setPrefBool("DisplayType", "Terminal", m_displayType);
}

void Terminal::setDisplayDateTime(bool enable) 
{
  m_displayDateTime = enable; 
  pSEQPrefs->setPrefBool("DisplayDateTime", "Terminal", m_displayDateTime);
}

void Terminal::setDisplayEQDateTime(bool enable) 
{
  m_displayEQDateTime = enable; 
  pSEQPrefs->setPrefBool("DisplayEQDateTime", "Terminal", m_displayEQDateTime);
}

void Terminal::setUseColor(bool enable)
{ 
  m_useColor = enable; 
  pSEQPrefs->setPrefBool("UseColor", "Terminal", m_useColor);
}

void Terminal::newMessage(const MessageEntry& message)
{
  MessageType type = message.type();

  // if the message type isn't enabled, nothing to do
  if ((((m_enabledTypes & ( uint64_t(1) << type)) == 0) &&
       ((m_enabledShowUserFilters & message.filterFlags()) == 0)) ||
      ((m_enabledHideUserFilters & message.filterFlags()) != 0))
    return;

  const char* setColor = 0;

  // if using color, and the type is within range, try to use color
  if ((m_useColor) && 
      (size_t(type) < (sizeof(typeStyles)/sizeof(const char*))))
  {
    // if a color is specified for the type, use it
    if (typeStyles[type])
    {
      setColor = typeStyles[type];
      fputs(setColor, stdout);
    }
  }
 
  QString text;

  // if displaying the type, add it
  if (m_displayType)
    text = MessageEntry::messageTypeString(type) + ": ";
    
  // if displaying the message date/time append it
  if (m_displayDateTime)
    text += message.dateTime().toString(m_dateTimeFormat) + " - ";

  // if displaying the messages eq date/time, append it
  if (m_displayEQDateTime && (message.eqDateTime().isValid()))
    text += message.eqDateTime().toString(m_eqDateTimeFormat) + " - ";

  // append the actual message text
  text += message.text();

  if (m_useColor)
    text.replace(m_itemPattern, 
		 QString("\e[1;4;34m\\2 (#\\1)\e[0;24m") + setColor);
  else
    text.replace(m_itemPattern, "\\2 (#\\1)");

  // now append the message text to the buffer
  fputs((const char*)text.latin1(), stdout);

  if (setColor)
    fputs(defaultStyle, stdout);
  
  fputc('\n', stdout);
  fflush (stdout);
}

#ifndef QMAKEBUILD
#include "terminal.moc"
#endif

